pkg/sources/github/feedlet.go

```diff
@@ -0,0 +1,396 @@
+	secretName = "secretName"
+	//// secretKey is the name of key inside the secret that contains the GitHub credentials.
+	secretKey = "secretKey"
+)
+
+type GithubEventSource struct {
+	// kubeclientset is a standard kubernetes clientset
+	kubeclientset kubernetes.Interface
+	// servingclientset is a clientset for serving API groups
+	servingclientset servingclientset.Interface
+	// namespace where the feed is created
+	feedNamespace string
+	// serviceAccount that the container runs as. Launches Receive Adapter with the
+	// same Service Account
+	feedServiceAccountName string
+	// image for the receive adapter
+	image string
+}
+
+func NewGithubEventSource(kubeclientset kubernetes.Interface, servingclientset servingclientset.Interface, feedNamespace, feedServiceAccountName, image string) sources.EventSource {
+	return &GithubEventSource{
+		kubeclientset:          kubeclientset,
+		servingclientset:       servingclientset,
+		feedNamespace:          feedNamespace,
+		feedServiceAccountName: feedServiceAccountName,
+		image:      image,
+	}
+}
+
+func (t *GithubEventSource) StopFeed(trigger sources.EventTrigger, feedContext sources.FeedContext) error {
+	glog.Infof("stopping github webhook feed with context %+v", feedContext)
+
+	serviceName := receiveAdapterName(trigger.Resource)
+	t.deleteReceiveAdapter(t.feedNamespace, serviceName)
+
+	components := strings.Split(trigger.Resource, "/")
+	owner := components[0]
+	repo := components[1]
+
+	if _, ok := feedContext.Context[webhookIDKey]; !ok {
+		// there's no webhook id, nothing to do.
+		glog.Error("no Webhook ID Found, bailing...")
+		return nil
+	}
+	webhookID := feedContext.Context[webhookIDKey].(string)
+
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: trigger.Parameters[accessToken].(string)},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+	client := ghclient.NewClient(tc)
+
+	id, err := strconv.ParseInt(webhookID, 10, 64)
+	if err != nil {
+		glog.Errorf("failed to convert webhook %q to int64 : %s", webhookID, err)
+		return err
+	}
+	_, err = client.Repositories.DeleteHook(ctx, owner, repo, id)
+	if err != nil {
+		if errResp, ok := err.(*ghclient.ErrorResponse); ok {
+			// If the webhook doesn't exist, nothing to do
+			if errResp.Message == "Not Found" {
+				glog.Errorf("webhook doesn't exist, nothing to delete.")
+				return nil
+			}
+		}
+		glog.Errorf("gailed to delete the webhook: %#v", err)
+		return err
+	}
+	glog.Infof("deleted webhook %q successfully", webhookID)
+	return nil
+}
+
+func (t *GithubEventSource) StartFeed(trigger sources.EventTrigger, target string) (*sources.FeedContext, error) {
+
+	// Create the Receive Adapter Service that will accept incoming requests from GitHub.
+	service, err := t.createReceiveAdapter(trigger, target)
+	if err != nil {
+		glog.Warningf("failed to create service: %v", err)
+		return nil, err
+	}
+
+	glog.Infof("created Service: %+v", service)
```

Comment:
We should consider using spew here. %v is pretty useless for most structs since it doesn't follow pointers.